tests/cxx/member_expr.cc

```diff
@@ -58,6 +58,65 @@ void ViaMacro(const IndirectClass& ic) {
 }
 
 
+// Test that referencing template specialization members needs full type info
```

Comment:
I'm having a hard time understanding this comment -- "[...] _all_ other members"? We only report anything for the member that was accessed, right? Or do you mean the template argument type?

Reply:
Member access constitutes full embracing type use, be it template specialization or just ordinary structure. Full use of template specialization type requires full type info for its components. It could be thought about as a compiler needs member sizes in order to compute the offset in bytes, but the standard and compilers require full type info for all members except functions, even for typedefs:
```c++
struct A;
struct B; // {};
struct C; // { typedef int InternalType; };

template <typename T1, typename T2, typename T3>
struct Template
{
  const T1& a;
  T2 b;
  typedef typename T3::InternalType CTypedef;
};

void fn(const Template<A, B, C>& t) // still compiles
{
  (void)t.a; // doesn't compile due to B and C; fwd-declaring A is still OK
}
```
And, of course, it concerns dependent types because non-dependent type info should already be available at the point of template definition.

Reply:
But if we had a template;
```
class A;

template<typename T>
struct Template {
    void method() const {}
};

void test(const Template<A>& t) {
   t.method();
}
``` 
that would not require the definition of `A`? It's only if there's a full-use member of template argument type in the template, then?

Just tested locally, and yeah, that appears to be the case. Adding a `T t;` member to `Template` causes it to no longer compile because it can't instantiate the template.

I think I was confused by the "referencing ... members ... needs full type info for all other members" phrasing. Would it be fair to say something like:
```
// Any member expression instantiates the template, check that the specialization requires all
// template type arguments that are used for members (data members, nested types) whether those
// types are part of the member expression or not.
```

(I don't think that's a great comment, but maybe could serve as inspiration :) )

Reply:
Very informative now, thanks!